includes/models/model.llms.membership.php

```diff
@@ -92,6 +93,65 @@ public function add_auto_enroll_courses( $course_ids, $replace = false ) {
 
 	}
 
+	/**
+	 * Retrieve a list of posts associated with the membership
+	 *
+	 * An associated post is:
+	 * + A post, page, or custom post type which supports `llms-membership-restrictions` and has restrictions enabled to this membership
+	 * + A course that exists in the memberships list of auto-enroll courses
+	 * + A course that has at least one access plan with members-only availability linked to this membership
+	 *
+	 * @since [version]
+	 *
+	 * @param string $post_type If supplied, returns only associations of this post type, otherwise returns an associative array of all associations.
+	 * @return array[]|int[] An array of arrays of post IDs. The array keys are the post type and the array values are arrays of integers.
+	 *                       If `$post_type` is supplied returns an array of associated post ids as integers.
+	 */
+	public function get_associated_posts( $post_type = null ) {
+
+		$posts = array();
+
+		// Retrieve all posts that are restricted to a membership via a LifterLMS Membership Restriction setting.
+		foreach ( get_post_types_by_support( 'llms-membership-restrictions' ) as $type ) {
+
+			// Skip if it's not the requested post type.
+			if ( $post_type && $type !== $post_type ) {
+				continue;
+			}
+
+			$posts[ $type ] = $this->query_associated_posts( $type, '_llms_is_restricted', 'yes', '_llms_restricted_levels' );
+
+		}
+
+		// Include courses if courses were requested or if no specific post type was requested.
+		if ( ! $post_type || 'course' === $post_type ) {
+
+			$posts['course'] = $this->query_associated_courses();
+
+		}
+
+		/**
+		 * Filter the list of posts associated with the membership.
+		 *
+		 * @since [version]
+		 *
+		 * @param array[]         $posts     An array of arrays of post IDs. The array keys are the post type and the array values are arrays of integers.
+		 * @param string          $post_type The requested post type if only a specific post type was requested, otherwise `null` to indicate all associated post types.
+		 * @param LLMS_Membership $this      Membership object.
+		 */
+		$posts = apply_filters( 'llms_membership_get_associated_posts', $posts, $post_type, $this );
+
+		// If a single post type was requested, return only that.
+		if ( $post_type ) {
+			// Return the request post type array and fallback to an empty array if that post type doesn't exist.
+			return isset( $posts[ $post_type ] ) ? $posts[ $post_type ] : array();
+		}
+
+		// Otherwise return everything.
+		return $posts;
```

Comment:
At a certain point I think we should remove the empty arrays, don't you think?
What's the meaning of having something like `array('course'=>array())` as result?

```suggestion
		return array_filter( $posts );
```

Reply:
I was thinking you could have an expected response that always has the keys so that you wouldn't have to always check the presence of a key before interacting the the response object. But now that I'm thinking about it I guess that wouldn't exactly matter because you couldn't always guarantee that it's going to be courses/posts/pages as a theme or plugin could modify post type support so you'd want to check the presence anwyay.

Reply:
I'll take a look at some modifications here and testse are going to have to be rewritten

Reply:
After reviwing I think your suggestion is perfect. I just updated the tests to accommodate this return.